language-server/vscoqtop/lspManager.ml

```diff
@@ -266,30 +270,62 @@ let textDocumentDidChange params =
         else 
           (st, [])
       in
-      Hashtbl.replace states (DocumentUri.to_path uri) st;
+      replace_state (DocumentUri.to_path uri) st visible;
       update_view uri st;
       inject_dm_events (uri, events)
 
 let textDocumentDidSave params =
   [] (* TODO handle properly *)
 
+let current_memory_usage () =
+  let { Gc.heap_words; _ } = Gc.stat () in
+  Sys.word_size * heap_words
+
+let purge_invisible_tabs () =
+  Hashtbl.filter_map_inplace (fun u ({ visible } as v) ->
+    if visible then Some v
+    else begin
+      log @@ "purging tab " ^ u;
+      None
+    end)
+  states
+
+let consider_purge_invisible_tabs () =
+  let usage = current_memory_usage () in
+  if usage > 4000000000 (* 4G *) then begin
+    purge_invisible_tabs ();
+    let vst, _ = get_init_state () in
+    Vernacstate.unfreeze_full_state vst;
+    Vernacstate.Interp.invalidate_cache ();
+    Gc.compact ();
+    let new_usage = current_memory_usage () in
+    log @@ Printf.sprintf  "memory footprint %d -> %d" usage new_usage;
+  end
+
 let textDocumentDidClose params =
+  let Lsp.Types.DidCloseTextDocumentParams.{ textDocument } = params in
+  let path = DocumentUri.to_path textDocument.uri in
+  begin match Hashtbl.find_opt states path with
+  | None -> assert false
+  | Some { st } -> replace_state path st false
```

Comment:
What's the reason for doing `replace` with `visible = false` then filtering instead of doing `Hashtbl.remove`?

Reply:
You don't necessarily want to remove the execution state of the document when you close it. For example you might close a document by mistake.

Reply:
The point is that, if memory allows for it, we want to keep the coq sessions alive.

It is not uncommon to close a tab and reopen it.
In VSCode you have have the usual browser shortcut for it, ctrl-w to close and shift-ctrl-t to open the last closed tab.
In this scenario we want to be nice to the users and not lose their session.

Reply:
Does reopening the tab get the saved state with the current code?

Also I guess the consider_purge should be delayed a bit since big sessions are the ones that are the worst to lose by mistake.

Reply:
Good question. Probably not at the moment.

For the consider_purge a reasonable alternative would be to send a popup message to the user when memory runs out asking him if he wants to purge ?